[For Review] HOTP and TOTP implementation

tekul opened the pull request:
This is the current implementation of the HOTP and TOTP specs I've been working on.

Any comments welcome, particularly on the API, use of config parameters etc.

tekul committed:
One-time password (OTP) implementation

Initial commit

- Implementation of HOTP algorithm as defined in RFC 4226
- Tests using values from the spec

tekul committed:
Add an OTP resynchronize function

Allows server to reset its counter to the client's current value, given
a sequence of one or more OTP values.

tekul committed:
Make OTP resynch values a tuple

This is clearer than having two separate arguments.

tekul committed:
Add TOTP function and KATs

Just uses SHA1 for now. HashAlgorithm is ignored.

tekul committed:
Add TOTParams data type

Reduce the arguments to the totp function (most people will use defaults)
and allows validation of the time step value.

Added a top-level module overview.

tekul committed:
Add hash parameter to hotp function

While HOTP only mentions SHA1, TOTP allows the use of different hash
functions, which implicitly requires that the HOTP implementation support
them too.

This will also allow users to use HOTP with another hash if they so choose,
though it would not be compatible with most client applications, such as
Google authenticator.

tekul committed:
Add TOTP KAT tests for SHA256 and SHA512

tekul committed:
Add totpVerify function

Also adds a ClockSkew type which limits the acceptable clock skew window
to a limited number of time steps.

vincenthz commented on Crypto/OTP.hs:
```diff
@@ -0,0 +1,168 @@
+    range n acc = range (n-1) ((t-n) : (t+n) : acc)
+
+timeToCounter :: POSIXTime -> Word64 -> Word16 -> Word64
+timeToCounter now t0 x = floor ((now - fromIntegral t0) / fromIntegral x)
```
POSIXTime is not a good representation. it has a 10^-12 representation that is useless here and I'm not a fan to depends on `time`, I think it would make more sense to just:

```
type OTPTime = Word64
or
type SecondSinceUnix = Word64
```

tekul replied:
It's true the extra resolution is no use and it's a pain to add a dependency just for this type, but presumably the end user will only be able to call `getPOSIXTime` or `fmap utcTimeToPOSIXSeconds getCurrentTime` to get the current time and would end up using this type in any case (they'd then have to convert). I don't know of any other functions which do the same thing. I guess we could add a function of type `IO OTPTime`  implemented in a similar way to these, but without needing the `time` dependency.

vincenthz replied:
They could (I would :-)) use `timeCurrent` from `hourglass` as well. I'm not sure it's a good idea to have a IO callback, as this would make the function impure, and un-replayable (if you want to test known values or passed/future values).

tekul replied:
Hmm, they could, yes. I forgot about that library :-). I meant just adding a function like `timeCurrent :: IO OTPTime` within the module itself, rather than modifying the function types as already written.

vincenthz commented on Crypto/OTP.hs:
```diff
@@ -0,0 +1,168 @@
+    -> Word64
+    -- ^ The T0 parameter in seconds. This is the Unix time from which to start
+    -- counting steps (default 0). Must be before the current time.
+    -> Word16
```
Int here maybe ? This is probably better as a more neutral Int than Word16, specially since it doesn't cover the whole Word16 range.

tekul replied:
Not sure what you mean - Word16 has a much smaller range than Int. If I used Int I'd also have to check for and reject negative values, so it seemed best to indicate in the type that it has to be positive.

tekul committed:
Drop use of 'time' library from OTP implementation

It now exposes a type alias for Word64 and relies on the user to supply
a value for the current time, allowing them to use the time library of
their choice.

Also bump memory dep to 0.12 and use fromW64BE from that library.
